sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -125,21 +125,23 @@ class Analyzer(
   object CTESubstitution extends Rule[LogicalPlan] {
     // TODO allow subquery to define CTE
     def apply(plan: LogicalPlan): LogicalPlan = plan resolveOperators  {
-      case With(child, relations) => substituteCTE(child, relations)
+      case With(child, relations) if relations.nonEmpty =>
+        var resolvedRelations = Seq(relations.head._1 -> ResolveRelations(relations.head._2))
+        relations.tail.foreach { case (name, relation) =>
+          resolvedRelations = resolvedRelations ++
+            Seq(name -> ResolveRelations(substituteCTE(relation, resolvedRelations)))
+        }
+        substituteCTE(child, resolvedRelations)
```

Comment:
Here, the CTEs of `WITH` clause are resolved sequentially before applying to main SQL body.